docs/zig-and-c.org

```diff
@@ -0,0 +1,249 @@
+You have better time importing the C code in your Zig code.
+Note that if you want to translate a C file that uses the libc you have to add the -lc flag like so:
+#+begin_src shell
+zig translate-c main.c -lc
+#+end_src
+
+**** Comparison with other langauges that use C code
+To test if integrating C code in Zig projects is really as seemless as some claims, I have decided to compare the C integration with Python aswell.
+
+In order to do that I wrote a small C library:
+#+begin_src c
+int add(int a, int b) { return a + b; }
+#+end_src
+
+What I am going to do is test how much time it takes each program to run this function x times. (in this case x = 100'000'000)
+
+Note: I didnt use any optimization in flag in python (because it did not change anything) and neither in Zig because I didnt want the compiler to try to optimize the code and be lazy.
+
+Then in order to compare the 2 languages I wrote 4 programs:
+- Zig code that has and add function implementation in Zig
+#+begin_src zig
+const std = @import("std");
+
+fn add(a: u32, b: u32) u32 {
+    return a + b;
+}
+
+pub fn main() !void {
+    var i: usize = 0;
+    while (i < 100000000) : (i += 1) {
+        _ = add(3, 7);
+    }
+    std.debug.print("done\n", .{});
+}
+#+end_src
+Result: ~0.38sec
+
+- Vanilla Python code that has and add function implementation in Python
+#+begin_src python
+def add(a, b):
+    return a + b
+
+
+for i in range(100000000):
+    add(3, 7)
+print("done!")
+#+end_src
+Result: ~10sec
+
+- Zig code that imports the C library
+#+begin_src zig
+const std = @import("std");
+pub const c = @cImport({
+    @cInclude("mylib.c");
+});
+
+pub fn main() !void {
+    var i: usize = 0;
+    while (i < 100000000) : (i += 1) {
+        _ = c.add(3, 7);
+    }
+    std.debug.print("done!\n", .{});
+}
+#+end_src
+Result: ~0.41sec
+
+- Python code that imports the C library
+#+begin_src python
+import ctypes
+
+mylib = ctypes.CDLL('./mylib.so')
+
+mylib.add.argtypes = (ctypes.c_int, ctypes.c_int)
+mylib.add.restype = ctypes.c_int
+
+for i in range(100000000):
+    result = mylib.add(3, 4)
+
+print("Result of last addition:", result)
+#+end_src
+Result: ~50sec
+
+TODO faire un joli graphique quand jaurai d autres languages
+TODO idk if it would be possible to compile my c library and then use it in my python code, and all of that all in emacs, that would be cool :)
+TODO do with other languages aswell
+
+***** Conclusion
+First thing that we notice immediately is how much faster the Zig code is compared to the Python code. This is not surprising since Zig is a compiled language and Python is an interpreted language.
+
+The second interesting thing is that the 2 Zig codes dont vary that much (if they even do) compared to the 2 python codes which have a 5x ratio. This is interesting because it shows that the overhead of calling a C function from Zig is not that big (in fact it is even not existent since all the compiler does is translating the C code to Zig code at compilation time).
```

Comment:
Was it the interoperability between Python and C that made the code slow or was it just the Python interpreter that couldn't optimise?

Reply:
very good question i added a TODO so that I can dig into it later :)